### PR TITLE
317 elastic scattering

### PR DESCRIPTION
**Pull request by satysh**
Доброго времени суток. Расширил ERElasticScattering. релятивистским случаем.
В ERElasticScattering.cxx перед имплемнтацией ERElasticScattering::ThetaRangesLab2CMRelativistic имеются комментарии, которые описывают что нового введено. Так же дополнил, в связи с этим, документацию.

### CODE REVIEW

File: decayers/ERElasticScattering.h [hunk 1]
```diff
@@ -140,6 +149,7 @@ class ERElasticScattering : public ERDecay
   Double_t        fTargetIonMass;             ///< Target ion mass
 
   Bool_t          fThetaLabRangeIsSet;        ///< kTRUE if Lab theta range is set, kFALSE if It is not set
+  Bool_t          fRelativisticMode;          ///< kTRUE of a relativistic case is simulated and kFALSE if It isn't
```
Comment (SchetininVitaliy) [inline decayers/ERElasticScattering.h]: Explicitly initialize attributes with proper value here and around. Bool_t fRelativisticMode = kFALSE; TF1* fThetaCDF = nullptr; Double_t        fProjectileIonMass = -1. ;   Double_t        fCDFmax = 1.;

File: decayers/ERElasticScattering.cxx [hunk 2]
```diff
@@ -103,16 +107,30 @@ Bool_t ERElasticScattering::Init() {
...
+    }
+  }
+  return kTRUE;
 }
 
 Bool_t ERElasticScattering::Stepping() {
+  if (fRelativisticMode && gMC->TrackPid() == fInputIonPDG->PdgCode()) {
+    ThetaRangesLab2CMRelativistic();
+    if (!ThetaCDFRead()) {
+      LOG(FATAL) << "The input file which contains the CDF function can't be read!" << FairLogger::endl;
```
Comment (SchetininVitaliy) [inline decayers/ERElasticScattering.cxx]: Move it to Init method

File: decayers/ERElasticScattering.cxx [hunk 3]
```diff
@@ -200,6 +217,14 @@ Bool_t ERElasticScattering::Stepping() {
       LOG(DEBUG) << "  Lab out2 state(px,py,pz,E) = "<<out2V.Px()<<","<<out2V.Py()<<","<<out2V.Pz()
                 << "," << out2V.E() << FairLogger::endl;
 
+      /* DEBUG*/
+      static Int_t nEvent = 0;
+      if (nEvent == 0) {
+        std::cerr << "[ Detector = " << setw(2) << fThetaRangeCenter << " ]:";
+        std::cerr << " thetaCM = " << generatedTheta*RadToDeg() << ", thetaLab = " << setw(9) << out2V.Theta()*RadToDeg();
```
Comment (SchetininVitaliy) [inline decayers/ERElasticScattering.cxx]: Move it to Init method. And you will not need this static staff.

File: decayers/ERElasticScattering.cxx [hunk 4]
```diff
@@ -103,16 +107,30 @@ Bool_t ERElasticScattering::Init() {
...
-  if (fThetaLabRangeIsSet)
+  if (!fRelativisticMode) {
     ThetaRangesLab2CM(GetProjectileIonMass(), GetTargetIonMass());
-
-  return ThetaCDFRead();
+    if (!ThetaCDFRead()) {
+      LOG(FATAL) << "The input file which contains the CDF function can't be read!" << FairLogger::endl;
+      return kFALSE;
+    }
+  }
```
Comment (SchetininVitaliy) [inline decayers/ERElasticScattering.cxx]: if (fRelativisticMode) {
   ThetaRangesLab2CMRelativistic()
} else {
  ThetaRangesLab2CM();
}
if (!ThetaCDFRead()) {
...

File: decayers/ERElasticScattering.cxx [hunk 5]
```diff
@@ -215,30 +240,31 @@ Bool_t ERElasticScattering::Stepping() {
...
 void  ERElasticScattering::ThetaRangesLab2CM(Double_t pM, Double_t tM) {
+  LOG(DEBUG) << "ERElasticScattering::ThetaRangesLab2CM(pM = " << pM << ", tM = " << tM << ")" << FairLogger::endl;
   Double_t rAng = fThetaRangeCenter*DegToRad();
   Double_t ratio = pM/tM;
   Double_t ratio2 = ratio*ratio;
-  Double_t rdTheta = fThetaRangedTheta*TMath::DegToRad(); // Detectors rdTheta
+  Double_t rdTheta = fThetaRangedTheta*DegToRad(); // Detectors rdTheta
   if (fRegisterIonStatus == kPROJECTILE) {
     // Projectile Ion
     if (pM != tM) {
```
Comment (SchetininVitaliy) [inline decayers/ERElasticScattering.cxx:226]: Please, explain, what does this condition mean?

File: decayers/ERElasticScattering.cxx [hunk 6]
```diff
@@ -215,30 +240,31 @@ Bool_t ERElasticScattering::Stepping() {
...
                 + cos(rAng+rdTheta)*sqrt(1.-ratio2*sin(rAng+rdTheta)*sin(rAng+rdTheta)) );
     }
     else {
-      fThetaMin = TMath::RadToDeg()*(2.*rAng - rdTheta);
-      fThetaMax = TMath::RadToDeg()*(2.*rAng + rdTheta);
+      fThetaMin = RadToDeg()*(2.*rAng - rdTheta);
+      fThetaMax = RadToDeg()*(2.*rAng + rdTheta);
     }
 
     LOG(DEBUG) << "  N15: CMTheta1: " << fThetaMin << ", CMTheta2: " << fThetaMax
```
Comment (SchetininVitaliy) [inline decayers/ERElasticScattering.cxx]: ERElasticScattering is a general class for reaction 2in2. Please do not specify it here

File: decayers/ERElasticScattering.cxx [hunk 7]
```diff
@@ -142,19 +160,18 @@ Bool_t ERElasticScattering::Stepping() {
       // Generate random angles theta and phi
       Double_t theta = ThetaGen();
       Double_t phi = fRnd->Uniform(fPhiMin*DegToRad(), fPhiMax*DegToRad());
-
+      Double_t generatedTheta = theta;
```
Comment (SchetininVitaliy) [inline decayers/ERElasticScattering.cxx]: You do not use this variable

File: decayers/ERElasticScattering.cxx [hunk 8]
```diff
@@ -215,30 +240,31 @@ Bool_t ERElasticScattering::Stepping() {
...
-  Double_t rdTheta = fThetaRangedTheta*TMath::DegToRad(); // Detectors rdTheta
+  Double_t rdTheta = fThetaRangedTheta*DegToRad(); // Detectors rdTheta
   if (fRegisterIonStatus == kPROJECTILE) {
     // Projectile Ion
     if (pM != tM) {
-      fThetaMin = TMath::RadToDeg()*acos( -ratio*sin(rAng-rdTheta)*sin(rAng-rdTheta)
+      fThetaMin = RadToDeg()*acos( -ratio*sin(rAng-rdTheta)*sin(rAng-rdTheta)
                 + cos(rAng-rdTheta)*sqrt(1.-ratio2*sin(rAng-rdTheta)*sin(rAng-rdTheta)) );
-      fThetaMax = TMath::RadToDeg()*acos( -ratio*sin(rAng+rdTheta)*sin(rAng+rdTheta)
+      fThetaMax = RadToDeg()*acos( -ratio*sin(rAng+rdTheta)*sin(rAng+rdTheta)
```
Comment (SchetininVitaliy) [inline decayers/ERElasticScattering.cxx]: Please, indicate in variable names where theta in Lab or CM

**Review by SchetininVitaliy — CHANGES_REQUESTED** [review]

File: decayers/ERElasticScattering.h (hunk 1 above)
Comment (satysh) [reply decayers/ERElasticScattering.h]: I didn't understand you clearly.
There are already initialized by default values. Initialization of all variables is implemented in the ERElastic*.cxx. In the implementation of the constructor **ERElasticScattering::ERElasticScattering(TString name)**

File: decayers/ERElasticScattering.cxx (hunk 2 above)
Comment (satysh) [reply decayers/ERElasticScattering.cxx]: I can't move It.
In this method ThetaRangesLab2CMRelativistic(); projectile momentum is used. 
To get this momentum I have to use 
 **TLorentzVector curLV;
  gMC->TrackMomentum(curLV);**

There is a problem that in Init() method gMC isn't accessible for use. 
The first point of access to gMC is in the Stepping() method.

I think, the solution that I chose for this problem isn't good. But unfortunately, I had only this ideas to solve It.
Can you give me some advises? How should I solve this problem?

File: decayers/ERElasticScattering.cxx (hunk 3 above)
Comment (satysh) [reply decayers/ERElasticScattering.cxx]: I think, we don't need this part. I am going to remove It.

File: decayers/ERElasticScattering.cxx (hunk 4 above)
Comment (satysh) [reply decayers/ERElasticScattering.cxx]: Good!

File: decayers/ERElasticScattering.cxx (hunk 5 above)
Comment (satysh) [reply decayers/ERElasticScattering.cxx:226]: This condition means that there are formulas inside this if(pM != tM) that work only for cases with pM != tM.
For the cases with pM =tM other formulas are used.

File: decayers/ERElasticScattering.cxx (hunk 6 above)
Comment (satysh) [reply decayers/ERElasticScattering.cxx]: Ok! I fixed It. I removed "N15" and wrote "Projectile".

File: decayers/ERElasticScattering.cxx (hunk 7 above)
Comment (satysh) [reply decayers/ERElasticScattering.cxx]: Ok! I removed It.

File: decayers/ERElasticScattering.cxx (hunk 8 above)
Comment (satysh) [reply decayers/ERElasticScattering.cxx]: Ok. I added Lab and CM in names of variables for corresponding thetas.

File: decayers/ERElasticScattering.h (hunk 1 above)
Comment (SchetininVitaliy) [reply decayers/ERElasticScattering.h]: In old c++ standard you need to initialize all variables in constructor using initialization list. But in new standards you can init variables right here in header.

Comment (satysh) [reply decayers/ERElasticScattering.h]: It is cool! I didn't know It. I will do It.

**Pull request merged**